drivers/SmartThings/matter-sensor/src/air-quality-sensor/init.lua

```diff
@@ -164,130 +164,101 @@ local supported_profiles =
-  end
+
+  local meas_name, level_name = create_level_measurement_profile(device)
+
   -- If all endpoints are supported, use '-all' in the profile name so that it
   -- remains under the profile name character limit
-  if level_indication_support == "-co-co2-no2-ozone-formaldehyde-pm1-pm25-pm10-radon-tvoc" then
-    level_indication_support = "-all"
+  if level_name == "-co-co2-no2-ozone-ch2o-pm1-pm25-pm10-radon-tvoc" then
+    level_name = "-all"
   end
-  if level_indication_support ~= "" then
-    profile_name = profile_name .. level_indication_support .. "-level"
+  if level_name ~= "" then
+    profile_name = profile_name .. level_name .. "-level"
   end
 
-  if #co_meas_eps > 0 then
-    numeric_measurement_support = numeric_measurement_support .. "-co"
-  end
-  if #co2_meas_eps > 0 then
-    numeric_measurement_support = numeric_measurement_support .. "-co2"
-  end
-  if #no2_meas_eps > 0 then
-    numeric_measurement_support = numeric_measurement_support .. "-no2"
-  end
-  if #ozone_meas_eps > 0 then
-    numeric_measurement_support = numeric_measurement_support .. "-ozone"
-  end
-  if #formaldehyde_meas_eps > 0 then
-    numeric_measurement_support = numeric_measurement_support .. "-formaldehyde"
-  end
-  if #pm1_meas_eps > 0 then
-    numeric_measurement_support = numeric_measurement_support .. "-pm1"
-  end
-  if #pm2_5_meas_eps > 0 then
-    numeric_measurement_support = numeric_measurement_support .. "-pm25"
-  end
-  if #pm10_meas_eps > 0 then
-    numeric_measurement_support = numeric_measurement_support .. "-pm10"
-  end
-  if #radon_meas_eps > 0 then
-    numeric_measurement_support = numeric_measurement_support .. "-radon"
-  end
-  if #tvoc_meas_eps > 0 then
-    numeric_measurement_support = numeric_measurement_support .. "-tvoc"
-  end
   -- If all endpoints are supported, use '-all' in the profile name so that it
   -- remains under the profile name character limit
-  if numeric_measurement_support == "-co-co2-no2-ozone-formaldehyde-pm1-pm25-pm10-radon-tvoc" then
-    numeric_measurement_support = "-all"
+  if meas_name == "-co-co2-no2-ozone-ch2o-pm1-pm25-pm10-radon-tvoc" then
+    meas_name = "-all"
   end
-  if numeric_measurement_support ~= "" then
-    profile_name = profile_name .. numeric_measurement_support .. "-meas"
+  if meas_name ~= "" then
+    profile_name = profile_name .. meas_name .. "-meas"
   end
 
   if not tbl_contains(supported_profiles, profile_name) then
     device.log.warn_with({hub_logs=true}, string.format("No matching profile for device. Tried to use profile %s", profile_name))
-    if #co_meas_eps > 0 or #no2_meas_eps > 0 or #ozone_meas_eps > 0 or #formaldehyde_meas_eps > 0 or
-        #pm1_meas_eps > 0 or #pm10_meas_eps > 0 or #radon_meas_eps > 0 then
+
+    local function meas_find(sub_name)
+      return string.match(meas_name, sub_name) ~= nil
+    end
+
+    -- try to best match to existing profiles
+    -- meas_find("co[^%d]?") matches the string to co and NOT co2.
```

Comment:
I know this isn't necessarily the scope of this PR, but since we are calling this out directly in the code it would be good to add a test case as an example of this working.

Reply:
Update: that suggested check `meas_find("co[^%d]?")` doesn't work unfortunately, since the `not` part doesn't reject something that's there, it just only returns what would be returned minus the part that is ^ marked. Thus, matching `co2-` would return `co`. I was playing with using `".*co[^%d].*"` as a match, but it fails when the string ends with co. And adding a ? after [^%d] in the second version makes -co2 pass. So basically lua matching functions are not complex enough to permit co but not co2 in all cases. Extra logic is needed.

Reply:
This pair of checks does work though to catch the case we're discussing: `meas_find("co%-") or meas_find("co$")`

Reply:
I'm glad we made some unit tests- good nose on a possible bug :+1: